src/repr/src/adt/mz_acl_item.rs

```diff
@@ -385,25 +361,110 @@ impl Columnation for MzAclItem {
+    pub fn from_mz_acl_items(items: impl IntoIterator<Item = MzAclItem>) -> PrivilegeMap {
+        let mut map = PrivilegeMap::new();
+        for item in items {
+            map.grant(item);
+        }
+        map
+    }
+
+    /// Get the acl item granted to `grantee` by `grantor`.
+    pub fn get_acl_item(&self, grantee: &RoleId, grantor: &RoleId) -> Option<&MzAclItem> {
+        self.0.get(grantee).and_then(|privileges| {
+            privileges
+                .into_iter()
+                .find(|mz_acl_item| &mz_acl_item.grantor == grantor)
+        })
+    }
+
+    /// Get all acl items granted to `grantee`.
+    pub fn get_acl_items_for_grantee(&self, grantee: &RoleId) -> impl Iterator<Item = &MzAclItem> {
+        self.0
+            .get(grantee)
+            .into_iter()
+            .flat_map(|privileges| privileges.into_iter())
     }
 
+    /// Returns references to all contained [`MzAclItem`].
     pub fn all_values(&self) -> impl Iterator<Item = &MzAclItem> {
         self.0
             .values()
             .flat_map(|privileges| privileges.into_iter())
     }
+
+    /// Returns clones of all contained [`MzAclItem`].
+    pub fn all_values_owned(&self) -> impl Iterator<Item = MzAclItem> + '_ {
+        self.all_values().cloned()
+    }
+
+    /// Adds an [`MzAclItem`] to this map.
+    pub fn grant(&mut self, privilege: MzAclItem) {
+        let grantee_privileges = self.0.entry(privilege.grantee).or_default();
+        if let Some(existing_privilege) = grantee_privileges
+            .iter_mut()
+            .find(|cur_privilege| cur_privilege.grantor == privilege.grantor)
+        {
+            // sanity check that the catalog data is consistent.
+            assert_eq!(
+                privilege.grantee, existing_privilege.grantee,
+                "catalog privileges out of sync"
+            );
+            existing_privilege.acl_mode = existing_privilege.acl_mode.union(privilege.acl_mode);
+        } else {
+            grantee_privileges.push(privilege);
+        }
+    }
+
+    /// Removes an [`MzAclItem`] from this map.
+    pub fn revoke(&mut self, privilege: &MzAclItem) {
+        let grantee_privileges = self.0.entry(privilege.grantee).or_default();
+        if let Some(existing_privilege) = grantee_privileges
+            .iter_mut()
+            .find(|cur_privilege| cur_privilege.grantor == privilege.grantor)
+        {
+            // sanity check that the catalog data is consistent.
+            assert_eq!(
+                privilege.grantee, existing_privilege.grantee,
+                "catalog privileges out of sync"
+            );
+            existing_privilege.acl_mode =
+                existing_privilege.acl_mode.difference(privilege.acl_mode);
+        }
+
+        // Remove empty privileges.
+        grantee_privileges.retain(|privilege| !privilege.acl_mode.is_empty());
+        if grantee_privileges.is_empty() {
+            self.0.remove(&privilege.grantee);
+        }
+    }
```

Comment:
These two methods are mostly code movement from `grant_object_privilege()` and `revoke_object_privilege()` from `catalog.rs`.